day8/alienalphabet.py

```diff
@@ -0,0 +1,131 @@
+##############
+## Solution ##
+##############
+
+def alienAlphabet(orderedWords):
+    # edge cases
+    if len(orderedWords) <= 0:
+        return ''
+    if len(orderedWords) == 1:
+        return removeDups(orderedWords[0])
+
+    # linearization of a DAG
+    graph = createDependencyGraph(orderedWords)
+    alphabet = topSort(graph)
+
+    return "".join(alphabet)
+
+#############
+## Helpers ##
+#############
+
+def createDependencyGraph(orderedWords):
+    graph = {}
+    for i, word in enumerate(orderedWords[:-1]):
+        nextWord = orderedWords[i+1]
+        shorter = min(len(word), len(nextWord))
+        for j in xrange(shorter):
+            letter = word[j]
+            otherLetter = nextWord[j]
+
+            if letter not in graph:
+                graph[letter] = []
+            if otherLetter not in graph:
+                graph[otherLetter] = []
```

Comment:
i was going to suggest defaultdicts, but i realized that you actually need to explicitly set the values for these keys because you depend on the assumption that every node is in the graph for `getSources`. in this case, you shouldn't be populating the graph inside of two nested for loops that have in a sense nothing to do with this. instead, at the beginning of the function you should've done this:

    graph = {}
    for word in orderedWords:
        for letter in word:
            graph[letter] = []

this makes it more clear what the keys of `graph` are, prevents you from accidentally overwriting a partially filled list (which you need `if` statements for here), and also allows you to do a really sick but kinda gross list comprehension like this instead of my above suggestion:

    graph = {letter: [] for word in orderedWords for letter in word}

notice the order of those two for loops. don't do it the other way around because it won't work